edgePy/DGEList.py

```diff
@@ -292,17 +292,40 @@ def rpkm(
         # self = self.cpm(log=log, prior_count=prior_count)
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
+            log: store log outputs
+            prior_count:
+            mean_fragment_lengths: vector of mean fragment length for each sample in counts table (optional)
 
         """
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
+                min=1
+            )
+        else:
+            effective_lengths = gene_lengths[:, None]
+
+        base_counts = self.counts / effective_lengths
+
+        self.counts = 10 ** 6 * base_counts / np.sum(base_counts, axis=0)[None, :]
+        if log:
```

Comment:
We don't assert `log` is a boolean so to be safe I always do:


```python
if log is True:
    ...
```

Otherwise other truthy values can slip in if the user is confused about what `log` is.

For example, this would be a valid value for `log` in this function:

```python
log = 1e10
```

Reply:
Hmm, I only partially agree. What do we have typings for (i.e. people have the documentation inside their IDEs stating, that it's a boolean. If somebody on purpose wants to fuck it up, he for sure can).

Reply:
Comes down to argument length vs readability so its up to the designer. I much prefer prepositions and verbs used to make a more readable expression at the sacrifice of a few more characters in typing. Examples for me are:


Conditional function: `maybe_do()`
Test membership: `has_object()`
Make a choice: `to_decide()`

With regard to typing. It is still a young feature in Python so used by developers using the library in their own. Many users of edgePy are likely to be working in Jupyter and type annotations are not going to help them.

If you keep argument name as a pure noun (`log`) it would be nice to have an early assertion validating it's type:

```python
if not isinstance(log, bool):
    raise TypeError('`log` must be of type bool.')
```

Reply:
I agree that naming the variables appropriately is the best way. I'll change the name in all three functions to 'transform_to_log' and see if anybody complains. 
Another possible name would be apply_log_transform(ation)